collection/collection.c

```diff
@@ -231,8 +231,9 @@ int col_allocate_item(struct collection_item **ci, const char *property,
     TRACE_INFO_NUMBER("Will be using type:", type);
 
     /* Check the length */
-    if (length >= COL_MAX_DATA) {
-        TRACE_ERROR_STRING("col_allocate_item", "Data to long.");
+    if ((length < 0) || (length >= COL_MAX_DATA) ||
+        ((type == COL_TYPE_STRING) && (length == 0))) {
```

Comment:
Perhaps, it should be `!= COL_TYPE_END`.

Reply:
Hi,

if `length` is `0` there might be issues later on with `malloc(length)`. But on the other hand it looks like `COL_TYPE_COLLECTION` is using `length == 0` as well, https://github.com/SSSD/ding-libs/blob/master/collection/collection.c#L3088. But after a short look at what is calling `col_allocate_item()` it looks like `COL_TYPE_COLLECTION` and similar types are handled differently and are not using `col_allocate_item()` at all.

To cut it short, maybe it would be easier to error out if `length < 1` unconditionally and print the type in the error message as well to be able to understand if `col_allocate_item()` was called with an unexpected type.

bye,
Sumit

Reply:
> if `length` is `0` there might be issues later on with `malloc(length)`. 

According to the man page, this is "legitimate".


> To cut it short, maybe it would be easier to error out if `length < 1` unconditionally and print the type in the error message

`COL_TYPE_END` uses 0:
https://github.com/SSSD/ding-libs/blob/master/collection/collection_iter.c#L104

Reply:
> > > if `length` is `0` there might be issues later on with `malloc(length)`.
> > 
> > According to the man page, this is "legitimate".
> 
> Yes, but `malloc()` is also allowed to return `NULL` in this case which would cause an `ENOMEM` error in the check after the `malloc()`. So I think the check after the `malloc()` should be a bit more elaborate to cover this ambiguous behavior if `length == 0`.

But if we allow `(item->data == NULL)` then we should protect `memcpy` because it's required to have valid ptrs even with `length == 0`

Please, see updated patch.